### DISCUSSION TITLE
Target Frameworks / Late in the year 2023.

### DISCUSSION THREAD

**Discussion by granadacoder** [thepirat000/Audit.NET#637] [Ideas]
I was working an a possible PR.

But wanted to discuss before doing the work.

The TargetFramework(s) .. are a little "all over the place"......
  I understand this "happens" with the long life of a .sln.

My primary concern (today) is :

Audit.NET.PostgreSql.csproj

does not have  net6.0 specified.

see:

https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.NET.PostgreSql/Audit.NET.PostgreSql.csproj#L7

But I see some discrepencies elsewhere.

IMHO, a late (year) 2023 list of appropriate target-frameworks is:

`

    <TargetFrameworks>netstandard2.0;netstandard2.1;net462;net6.0;net7.0</TargetFrameworks>

`

based on:

https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core

dot-net-CORE:

net6.0 because it is the latest LTS version.  (and what my company writes to)
and net7.0 because it is the current latest/greatest STS version...even though net8.0 is knocking on the door.

The "debatables" for dot-net CORE.

netstandard1.3 (or anything pre 2.0)
and
netcoreapp3.1 (but my opinion on 3.1 is now "remove it")

and

Dotnet-FRAMEWORK

https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework

.NET Framework 4.6.2 | Aug 2, 2016 | Jan 12, 2027

anything "pre" 4.6.2 should be removed.

Let me know your thoughts before I take on too-little/too-less.

I can do the "minimal" (add net6.0 to the postgres csproj) .. or do a little more.

Note, in the below, the isolated "8" or "7" (etc, etc) numbers after ".csproj" are the **_line_** numbers where the TargetFrameworks are defined in the csproj.

`

    Occurrences of '</TargetFrameworks>

        <Audit.DynamicProxy> 
            Audit.DynamicProxy.csproj 
                8 <TargetFrameworks>netstandard1.5;net45;net461</TargetFrameworks>

        <Audit.EntityFramework.Core.v3> 
            Audit.EntityFramework.Core.v3.csproj 
                7 <TargetFrameworks>netstandard2.0;net5.0</TargetFrameworks>

        <Audit.EntityFramework.Core> 
            Audit.EntityFramework.Core.csproj 
                8 <TargetFrameworks>netstandard1.5;netstandard2.0;net461;netstandard2.1;net472;net5.0;net6.0;net7.0</TargetFrameworks>

        <Audit.EntityFramework.Identity.Core.v3> 
            Audit.EntityFramework.Identity.Core.v3.csproj 
                7 <TargetFrameworks>netstandard2.1;net5.0</TargetFrameworks>

        <Audit.EntityFramework.Identity.Core> 
            Audit.EntityFramework.Identity.Core.csproj 
                8 <TargetFrameworks>netstandard1.5;netstandard2.1;net461;net5.0;net6.0;net7.0</TargetFrameworks>

        <Audit.EntityFramework.Identity> 
            Audit.EntityFramework.Identity.csproj 
                8 <TargetFrameworks>netstandard1.5;netstandard2.0;net45;netstandard2.1;net5.0</TargetFrameworks>

        <Audit.EntityFramework> 
            Audit.EntityFramework.csproj 
                8 <TargetFrameworks>netstandard1.5;netstandard2.0;net45;netstandard2.1;net472;net5.0</TargetFrameworks>

        <Audit.FileSystem> 
            Audit.FileSystem.csproj 
                8 <TargetFrameworks>netstandard1.3;net45;net5.0</TargetFrameworks>

        <Audit.HttpClient> 
            Audit.HttpClient.csproj 
                7 <TargetFrameworks>netstandard1.3;netstandard2.0;netstandard2.1;net45;net461;net5.0</TargetFrameworks>

        <Audit.Mvc.Core> 
            Audit.Mvc.Core.csproj 
                8 <TargetFrameworks>netstandard1.6;netstandard2.0;net451;netstandard2.1;netcoreapp3.0;net5.0</TargetFrameworks>

        <Audit.Mvc> 
            Audit.Mvc.csproj 
                8 <TargetFrameworks>netstandard1.6;netstandard2.0;net45;netstandard2.1;netcoreapp3.0;net5.0</TargetFrameworks>

        <Audit.NET.AzureCosmos> 
            Audit.NET.AzureCosmos.csproj 
                9 <TargetFrameworks>net45;netstandard1.6;netstandard2.0;net5.0</TargetFrameworks>

        <Audit.NET.AzureStorage> 
            Audit.NET.AzureStorage.csproj 
                8 <TargetFrameworks>netstandard1.4;net452;net5.0</TargetFrameworks>

        <Audit.NET.AzureStorageBlobs> 
            Audit.NET.AzureStorageBlobs.csproj 
                8 <TargetFrameworks>netstandard2.0;net5.0</TargetFrameworks>

        <Audit.NET.AzureStorageTables> 
            Audit.NET.AzureStorageTables.csproj 
                8 <TargetFrameworks>netstandard2.0;net5.0</TargetFrameworks>

        <Audit.NET.DynamoDB> 
            Audit.NET.DynamoDB.csproj 
                8 <TargetFrameworks>netstandard2.0;net45;net5.0</TargetFrameworks>

        <Audit.NET.Elasticsearch> 
            Audit.NET.Elasticsearch.csproj 
                8 <TargetFrameworks>net461;netstandard2.0</TargetFrameworks>

        <Audit.NET.EventLog.Core> 
            Audit.NET.EventLog.Core.csproj 
                8 <TargetFrameworks>netstandard2.0;net5.0</TargetFrameworks>

        <Audit.NET.JsonNewtonsoftAdapter> 
            Audit.NET.JsonNewtonsoftAdapter.csproj 
                7 <TargetFrameworks>net5.0</TargetFrameworks>

        <Audit.NET.JsonSystemAdapter> 
            Audit.NET.JsonSystemAdapter.csproj 
                7 <TargetFrameworks>netstandard2.0;net461</TargetFrameworks>

        <Audit.NET.Kafka> 
            Audit.NET.Kafka.csproj 
                7 <TargetFrameworks>netstandard2.0;netstandard1.3;net45;net5.0</TargetFrameworks>

        <Audit.NET.MongoDB> 
            Audit.NET.MongoDB.csproj 
                8 <TargetFrameworks>netstandard1.5;net452;net472;netstandard2.0;net5.0</TargetFrameworks>

        <Audit.NET.MySql> 
            Audit.NET.MySql.csproj 
                8 <TargetFrameworks>netstandard1.3;netstandard2.0;netcoreapp2.1;net45;net461;net5.0</TargetFrameworks>

        <Audit.NET.NLog> 
            Audit.NET.NLog.csproj 
                9 <TargetFrameworks>netstandard1.3;net45;net5.0</TargetFrameworks>

        <Audit.NET.PostgreSql> 
            Audit.NET.PostgreSql.csproj 
                7 <TargetFrameworks>netstandard2.0;netstandard2.1;net461;net6.0</TargetFrameworks>

        <Audit.NET.RavenDB> 
            Audit.NET.RavenDB.csproj 
                9 <TargetFrameworks>net5.0;net6.0;netstandard2.0</TargetFrameworks>

        <Audit.NET.Redis> 
            Audit.NET.Redis.csproj 
                8 <TargetFrameworks>netstandard2.0;net461;net5.0</TargetFrameworks>

        <Audit.NET.Serilog> 
            Audit.NET.Serilog.csproj 
                8 <TargetFrameworks>netstandard1.3;net45;net5.0</TargetFrameworks>

        <Audit.NET.SqlServer> 
            Audit.NET.SqlServer.csproj 
                8 <TargetFrameworks>netstandard2.1;netstandard2.0;netstandard1.3;net45;net462;net5.0;net6.0;net7.0</TargetFrameworks>

        <Audit.NET.Udp> 
            Audit.NET.Udp.csproj 
                8 <TargetFrameworks>netstandard1.3;net45</TargetFrameworks>

        <Audit.NET.log4net> 
            Audit.NET.log4net.csproj 
                8 <TargetFrameworks>netstandard1.3;net45</TargetFrameworks>

        <Audit.NET> 
            Audit.NET.csproj 
                7 <TargetFrameworks>netstandard1.3;netstandard2.0;netstandard2.1;net45;net461;net5.0;net6.0;net7.0</TargetFrameworks>

        <Audit.SignalR> 
            Audit.SignalR.csproj 
                7 <TargetFrameworks>net45;net7.0</TargetFrameworks>

        <Audit.WCF.Client> 
            Audit.Wcf.Client.csproj 
                7 <TargetFrameworks>net45;netstandard2.0</TargetFrameworks>

        <Audit.WebApi.Core> 
            Audit.WebApi.Core.csproj 
                8 <TargetFrameworks>netstandard1.6;netstandard2.0;net451;netstandard2.1;netcoreapp3.0;net5.0</TargetFrameworks>

        <Audit.WebApi> 
            Audit.WebApi.csproj 
                8 <TargetFrameworks>netstandard1.6;netstandard2.0;net45;netstandard2.1;netcoreapp3.0;net461;net5.0</TargetFrameworks>

    <test>  (14 usages found)
        <Audit.AzureStorageTables.UnitTest> 
            Audit.AzureStorageTables.UnitTest.csproj 
                4 <TargetFrameworks>netcoreapp2.1;net5.0</TargetFrameworks>

        <Audit.DynamicProxy.UnitTest> 
            Audit.DynamicProxy.UnitTest.csproj 
                5 <TargetFrameworks>netcoreapp1.0;net451;net461;net5.0</TargetFrameworks>

        <Audit.EntityFramework.Core.UnitTest> 
            Audit.EntityFramework.Core.UnitTest.csproj 
                5 <TargetFrameworks>netcoreapp2.1;netcoreapp3.0;net472;net5.0;net461;netcoreapp1.0;net6.0;net7.0</TargetFrameworks>

        <Audit.EntityFramework.Core.v3.UnitTest> 
            Audit.EntityFramework.Core.v3.UnitTest.csproj 
                5 <TargetFrameworks>netcoreapp2.1;netcoreapp3.0;net472;net5.0</TargetFrameworks>

        <Audit.EntityFramework.Full.UnitTest> 
            Audit.EntityFramework.Full.UnitTest.csproj 
                5 <TargetFrameworks>netcoreapp3.0;net472;net5.0</TargetFrameworks>

        <Audit.IntegrationTest> 
            Audit.IntegrationTest.csproj 
                5 <TargetFrameworks>net452;netcoreapp2.0;net461;netcoreapp3.0;net5.0;net6.0;net7.0</TargetFrameworks>

        <Audit.JsonAdapter.UnitTest> 
            Audit.JsonAdapter.UnitTest.csproj 
                4 <TargetFrameworks>netcoreapp2.0;net461;net5.0;net6.0</TargetFrameworks>

        <Audit.Mvc.UnitTest> 
            Audit.Mvc.UnitTest.csproj 
                5 <TargetFrameworks>netcoreapp1.0;netcoreapp2.0;net451;net45;net461;netcoreapp3.1;net5.0</TargetFrameworks>

        <Audit.RavenDB.UnitTest> 
            Audit.RavenDB.UnitTest.csproj 
                4 <TargetFrameworks>netcoreapp3.1;net5.0;net6.0</TargetFrameworks>

        <Audit.Redis.UnitTest> 
            Audit.Redis.UnitTest.csproj 
                5 <TargetFrameworks>netcoreapp3.0;net461;net5.0</TargetFrameworks>

        <Audit.SqlServer.UnitTest> 
            Audit.SqlServer.UnitTest.csproj 
                4 <TargetFrameworks>net45;net462;netcoreapp2.0;netcoreapp3.0;net5.0;net6.0;net7.0</TargetFrameworks>

        <Audit.UnitTest> 
            Audit.UnitTest.csproj 
                5 <TargetFrameworks>netcoreapp3.0;netcoreapp2.0;netcoreapp1.0;net461;net5.0;net6.0;net7.0</TargetFrameworks>

        <Audit.Wcf.UnitTest> 
            Audit.Wcf.UnitTest.csproj 
                5 <TargetFrameworks>net45;net6.0</TargetFrameworks>

        <Audit.WebApi.UnitTest> 
            Audit.WebApi.UnitTest.csproj 
                5 <TargetFrameworks>netcoreapp1.0;netcoreapp2.0;net451;net45;net461;netcoreapp3.1;net5.0</TargetFrameworks>

`

**Comment by thepirat000** [comment]
I like the idea.

We'll be losing compatibility, anyone targeting an older framework won't be able to update to the latest version of the library... But I think it's the way to go.

Note you should also search for `</TargetFramework>` since there are some project that only targets one framework

**Comment by thepirat000** [comment]
This has been completed in version 23.0.0

Now the library and all its extensions target net462, netstandard2.0 and/or net6.0 as the minimum target framework. Older framework support has been deprecated. 

More information [here](https://github.com/thepirat000/Audit.NET?tab=readme-ov-file#important-note---support-for-older-net-frameworks)

This also resulted in a lot of code clean-up, unit test separation, removal of preprocessor directives, etc. 

Note that `Audit.NET.PostgreSql` 23.0.0 now also target net6.0
